spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -220,6 +232,9 @@ private void addChannelFromString(Collection<MessageChannel> channels, String ch
 		MessageChannel channel = resolveChannelForName(channelName, message);
 		if (channel != null) {
 			channels.add(channel);
+			if (!mapped && !this.dynamicChannels.contains(channelName)) {
+				this.dynamicChannels.add(channelName);
```

Comment:
Well, "weak references" sounds like "eviction policy" and that doesn't fit your "it should remain" rule.
If we follow with your point, we should leave everything as is.

But... Thinking twice, even if we leave if we won't have the link for the dynamic channel. We have a protection I see:
```
channelNode = channelNodes.get(route);
if (channelNode != null) {
...
}
```
So, if that would be enough easy to turn it into weak reference, feel free to go ahead. Otherwise let's just live with the `Set<String>`!